consul/src/main/java/com/linecorp/armeria/internal/consul/ConsulClientBuilder.java

```diff
@@ -0,0 +1,151 @@
+        requireNonNull(consulUri, "consulUri");
+        final String path = consulUri.getPath();
+        checkArgument(!Strings.isNullOrEmpty(path) && !"/".equals(path),
+                      "consulUri has to contain version path");
+        this.consulUri = consulUri;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's API service URI.
+     * The URI should include the Consul API version at path, like: {@code /v1}
+     * @param consulUri the URI of Consul API service, default: {@code http://127.0.0.1:8500/v1}
+     */
+    public ConsulClientBuilder consulUri(String consulUri) {
+        return consulUri(URI.create(requireNonNull(consulUri, "consulUri")));
+    }
+
+    /**
+     * Sets the specified Consul's API service protocol scheme.
+     * @param consulProtocol the protocol scheme of Consul API service, default: {@link SessionProtocol#HTTP}
+     */
+    public ConsulClientBuilder consulProtocol(SessionProtocol consulProtocol) {
+        checkState(consulUri == null, "consulProtocol can't comes with consulUri");
+        requireNonNull(consulProtocol, "consulProtocol");
+        this.consulProtocol = consulProtocol;
+        isUriComponentSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's API service host address.
+     * @param consulAddress the host address of Consul API service, default: {@code 127.0.0.1}
+     */
+    public ConsulClientBuilder consulAddress(String consulAddress) {
+        checkState(consulUri == null, "consulAddress can't comes with consulUri");
+        requireNonNull(consulAddress, "consulAddress");
+        checkArgument(!consulAddress.isEmpty(), "consulAddress can't be empty");
+        this.consulAddress = consulAddress;
+        isUriComponentSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's HTTP service port.
+     * @param consulPort the port of Consul agent, default: {@code 8500}
+     */
+    public ConsulClientBuilder consulPort(int consulPort) {
+        checkState(consulUri == null, "consulPort can't comes with consulUri");
+        checkArgument(consulPort > 0, "consulPort can't be zero or negative");
+        this.consulPort = consulPort;
+        isUriComponentSet = true;
+        return this;
+    }
+
+    /**
+     * Sets the specified Consul's API version.
+     * @param consulApiVersion the version of Consul API service, default: {@code /v1}
+     */
+    public ConsulClientBuilder consulApiVersion(String consulApiVersion) {
+        checkState(consulUri == null, "consulApiVersion can't comes with consulUri");
+        requireNonNull(consulApiVersion, "consulApiVersion");
+        checkArgument(!consulApiVersion.isEmpty(), "consulApiVersion can't be empty");
+        checkArgument(consulApiVersion.charAt(0) != '/',
+                      "consulApiVersion can't starts with '/'");
+        this.consulApiVersion = consulApiVersion;
+        isUriComponentSet = true;
+        return this;
+    }
```

Comment:
Do you think it's a good idea to remove all these methods in favor of `consulUri`? Consul agent seems to run at localhost in most cases, so I guess these methods will not be used very often.

Reply:
Does that mean it would be better to remove the alternative methods except for consul Uri?
These methods are designed for such cases, but if they are verbose we can safely delete them. However, it may be confusing to users regarding the version. Most users will not be interested in the API version. Would it be better to leave only the URI and version?

Reply:
Nonetheless, I think it wouldn't be bad to leave it as it is. Because in many cases we will only change the port or protocol, like this:
```
ConsulEndpointGroup endpointGroup =
                     ConsulEndpointGroup.builder(SERVICE_NAME)
                                        .consulPort(CONSUL_AGENT_PORT)
                                        .consulToken(TOKEN)
                                        .build();
```

Reply:
OK. That's fine as well.